### PR TITLE
Feature/GNSE_LPM

### LINKED ISSUE #142: Support SE OTAA joins and operations in low power mode
**Summary:** 

Need to add support for secure element LoRaWAN OTAA joins and operations in low power mode.

**Why do we need this?**

To be able to conserve energy when the secure element is operational.

**What is already there? What do you see now?**

The device joins OTAA in active power mode

**What is missing? What do you want to see?**

The device operates and joins in low power mode

**How do you propose to implement this?**

Save device Temp Data in register to allow for re-utilization after device wake-up

**Environment:**

Bare-metal

**Acceptance Criteria:**

...

**What can you do yourself and what do you need help with?**

All

### LINKED ISSUE #171: Improve GNSE LPM to support board peripherals
**Summary:** 

We need to add a smart low power manager (LPM) that allows the system to enter stop mode and save power when multiple onboard components are active.

**Why do we need this?**

To be able to have longer battery life with multiple sensors.

**What is already there? What do you see now?**

LPM that supports UART sleep.

**What is missing? What do you want to see?**

An LPM that supports, sensors, buzzers and secure element.

**How do you propose to implement this?**

By Improving current LPM to detect HAL inits and support turning on and off the correct peripherals.

**Environment:**

Bare-metal and OS based

**Acceptance Criteria:**

If sensors applications sleep with reduced power consumption.
All applications compile and perform as expected.

**What can you do yourself and what do you need help with?**

All

### PR DESCRIPTION
**Pull request by elsalahy**
**Summary:**

This adds an improved low power manager that allows applications to reduce power consumption significantly.
The LPM also provides multiple APIs and Reduces the amount of configuration needed to reach low power.
Closes #171 
Closes #142 

**Changes:**

- Removed deprecated underlying utility `STM32_lpm.h`
- Added multipled APIs to GNSE_lpm.h to handle device peripherals
- Updated applications and config to use new GNSE_lpm.h

**Notes for Reviewers:**

- To reach lowest power consumption, disable the debugger and tracer using the conf file.
- `freefall_lorawan` needs more optimization and we have #180 for this.
- This app could break things, so please avoid nitpicking and please focus on functionally and API changes

**Release Notes: _(optional)_**

...

### CODE REVIEW

**Comment by elsalahy** [comment]
@marnixcro it would be appreciated if you could test the entire functionality of the following applications in your testing
- `freefall_lorawan`
- `sensors_lorawan`

I already tested
- `basic_lorawan`
- `secure_element_lorawan`

> also all applications compile using CMAKE and STM32CubeIDE

**Review by marnixcro — APPROVED** [review]
I could only find nitpicks. All functionalities I tested seemed to work:
- Compiling works on STM32CubeIDE and CMake
- basic app functionality seems to work (sensors, lorawan, etc.)
- freefall_lorawan's sensor load switch stayed on during lpm
- UART works
- Debugger works
The baremetal applications (except freefall_lorawan) are all <5uA, most under this

**Comment by elsalahy** [comment]
@marnixcro awesome, I also took care of the nitpicks 🚀

File: Software/app/freefall_lorawan/README.md [hunk 1]
```diff
@@ -19,8 +19,8 @@ This application is used to detect free fall events and also serves as a templat
 
 ### Low power
 
-1. Make sure that `LOW_POWER_DISABLE` is kept at `0` in [`conf/app_conf.h`](./conf/app_conf.h).
-2. Set `GNSE_ADVANCED_TRACER_ENALBE` to `0` in [`conf/app_conf.h`](./conf/app_conf.h). This will make the device consume less power, but also disables UART.
+1. Set `GNSE_ADVANCED_TRACER_ENALBE` to `0` in [`conf/app_conf.h`](./conf/app_conf.h). This will make the device consume less power, but also disables UART.
+2. Make sure that `DEBUGGER_ON` is kept at `0` in [`conf/app_conf.h`](./conf/app_conf.h).
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/freefall_lorawan/README.md]: Fixed

File: Software/lib/GNSE_HAL/GNSE_lpm.c [hunk 2]
```diff
@@ -21,98 +21,229 @@
...
+}
 
 /**
-  * @brief Enters Low Power Off Mode
-  * @param none
-  * @return none
-  */
-void GNSE_LPM_EnterOffMode(void)
+ * @brief This function will be called before entering Sleep Mode
+ * @note This API can be used to do an application task before enetering Sleep Mode
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/lib/GNSE_HAL/GNSE_lpm.c]: Fixed

**Review by marnixcro — APPROVED** [review]

**Pull request merged**